api/verify.go

```diff
@@ -390,3 +312,81 @@ func (a *API) emailChangeVerify(ctx context.Context, conn *storage.Connection, p
+		switch params.Type {
+		case signupVerification, inviteVerification:
+			user, err = models.FindUserByConfirmationToken(conn, params.Token)
+		case recoveryVerification, magicLinkVerification:
+			user, err = models.FindUserByRecoveryToken(conn, params.Token)
+		case emailChangeVerification:
+			user, err = models.FindUserByEmailChangeToken(conn, params.Token)
+		}
+	} else if params.Type == smsVerification {
+		if params.Phone == "" {
+			return nil, unprocessableEntityError("Sms Verification requires a phone number")
+		}
+		params.Phone = a.formatPhoneNumber(params.Phone)
+		if ok := a.validateE164Format(params.Phone); !ok {
+			return nil, unprocessableEntityError("Invalid phone number format")
+		}
+		user, err = models.FindUserByPhoneAndAudience(conn, instanceID, params.Phone, aud)
+	} else if params.Email != "" {
+		if err := a.validateEmail(ctx, params.Email); err != nil {
+			return nil, unprocessableEntityError("Invalid email format").WithInternalError(err)
+		}
+		user, err = models.FindUserByEmailAndAudience(conn, instanceID, params.Email, aud)
+	}
+
+	if err != nil {
+		if models.IsNotFoundError(err) {
+			return nil, notFoundError(err.Error()).WithInternalError(redirectWithQueryError)
+		}
+		return nil, internalServerError("Database error finding user").WithInternalError(err)
+	}
+
+	if user.IsBanned() {
+		return nil, unauthorizedError("Error confirming user").WithInternalError(redirectWithQueryError)
+	}
+
+	var isValid bool
+	mailerOtpExpiresAt := time.Second * time.Duration(config.Mailer.OtpExp)
+	smsOtpExpiresAt := time.Second * time.Duration(config.Sms.OtpExp)
+	switch params.Type {
+	case signupVerification, inviteVerification:
+		isValid = isOtpValid(params.Token, user.ConfirmationToken, user.ConfirmationSentAt.Add(mailerOtpExpiresAt))
+	case recoveryVerification, magicLinkVerification:
+		isValid = isOtpValid(params.Token, user.RecoveryToken, user.RecoverySentAt.Add(mailerOtpExpiresAt))
+	case emailChangeVerification:
+		expiresAt := user.EmailChangeSentAt.Add(mailerOtpExpiresAt)
+		isValid = isOtpValid(params.Token, user.EmailChangeTokenCurrent, expiresAt) || isOtpValid(params.Token, user.EmailChangeTokenNew, expiresAt)
+		if !isValid {
+			// reset email confirmation status
+			user.EmailChangeConfirmStatus = zeroConfirmation
+			err = conn.UpdateOnly(user, "email_change_confirm_status")
+		}
+	case smsVerification:
+		isValid = isOtpValid(params.Token, user.ConfirmationToken, user.ConfirmationSentAt.Add(smsOtpExpiresAt))
+	}
+
+	if !isValid || err != nil {
+		return nil, expiredTokenError("Token has expired or is invalid").WithInternalError(redirectWithQueryError)
+	}
+	return user, nil
+}
+
+// isOtpValid checks the actual otp sent against the expected otp and ensures that it's within the valid window
+func isOtpValid(actual, expected string, expiresAt time.Time) bool {
+	return time.Now().Before(expiresAt) && (actual == expected)
+}
+
+func isUrlVerification(params *VerifyParams) bool {
```

Comment:
General comment:  this seems to check by excluding(e.g. if it's not an sms verification or email verification then it is a URLVerification). Would be nice if we could do something like `params.Type == URLVerification`. That said, it would require a refactor and is probably not within a scope of this PR.

Reply:
Yeah i thought about that too but the type (signup vs recover) is tightly coupled to the `ConfirmationToken` & `RecoveryToken` fields, so it would be hard to tell whether the URL is a confirmation or recovery URL if we introduced a `params.Type == URLVerification`